source/language/scope.rst

```diff
@@ -0,0 +1,318 @@
+
+   defcal h $0 {
+      // ...
+   }
+   // No error: this is valid OpenQASM 3.0 because `defcal` statements do not
+   // redefine, they overload quantum operations with specific pulse-level
+   // control statements.
+
+   defcal a $0 {
+      // ...
+   }
+   // ERROR: 'a' is already defined and cannot be re-declared.  Unlike the
+   // previous example, this is an error because 'a' is already declared as a
+   // non-quantum-operation type ('uint').  This `defcal` would be defining a
+   // new gate, which is invalid with 'a' already defined.
+
+
+.. _scope-subroutine:
+
+Subroutine and gate scope
+-------------------------
+
+The definitions of subroutines (``def``) and gates (``gate``) introduce a new
+scope.  The ``def`` and ``gate`` statements are only valid directly within the
+global scope of the program.
+
+Inside the definition of the subroutine or gate, symbols that were already
+defined in the global scope with the ``const`` modifier, or previously defined
+gates and subroutines are visible.  Globally scoped variables without the
+``const`` modifier are not visible inside the definition.  In other words,
+subroutines and gates cannot close over variables that may be modified at
+run-time.
+
+Variables defined in subroutine scopes are local to the subroutine body.
+Variables defined in the parameter specifications of subroutines and gates
+behave for scoping purposes as if they were defined in the scope of the
+definition.  The lifetime of these local variables ends at the end of the
+function body, and they are not accessible after the subroutine or gate body.
+Similarly, the qubit identifiers in a gate definition are valid only within the
+definition of the gate.
+
+The identifier of a subroutine or gate is available in the scope of its own
+body, allowing direct recursion.  For gates, the direct recursion is unlikely to
+ever be useful, since this would generally be non-terminating.
+
+Local subroutine or gate variables, including parameters and qubit definitions,
+may shadow variables defined in the outer scope.  Inside the body, the
+identifier will refer to the local variable instead.  After the definition of
+the body has completed (and we are back in the global scope), the identifier
+will refer to the same variable it did before the subroutine or gate.
+
+Subroutines cannot contain ``qubit`` declarations in their bodies, but can
+accept variables of type ``qubit`` in their parameter lists.  Aliases can be
+declared within subroutine and gate scopes, and have the same lifetime and
+visibility as other local variables.
+
+For example:
+
+.. code-block:: c
+   :linenos:
+
+   OPENQASM 3.0;
+
+   qubit[5] all_qubits;
+
+   int a = 1;
+   int b = 2;
+   const int c = 3;
+   const int d = 4;
+
+   def my_routine(uint a, uint c) {
+      // In this body, 'a' refers to the subroutine parameter, not the external
+      // variable, which wouldn't be visible even without the shadowing.
+
+      int local = 5;
```

Comment:
**nitpick**: Use a different name than `local` since the word already carries a meaning in the discussion. Same deal on line 223.

Reply:
Good idea - done in dbc451c.